tests/unit/base/SGObject_unittest.cc

```diff
@@ -586,3 +588,195 @@ TEST(SGObjectAll, DISABLED_tag_coverage)
+template <typename T>
+class sg_object_iterator
+{
+public:
+	sg_object_iterator()
+	{
+		m_class_names = available_objects();
+	}
+
+	sg_object_iterator(std::set<std::string> ignores)
+	{
+		m_class_names = available_objects();
+
+		std::set<std::string> diff;
+
+		std::set_difference(
+		    m_class_names.begin(), m_class_names.end(), ignores.begin(),
+		    ignores.end(), std::inserter(diff, diff.begin()));
+		m_class_names = diff;
+	}
+
+	class Iterator : public std::iterator<std::input_iterator_tag, T>
+	{
+	public:
+		Iterator(
+		    std::set<std::string>::iterator it,
+		    std::set<std::string>::iterator end)
+		    : m_it(it), m_end(end), m_obj(nullptr)
+		{
+			create_and_forward_if_necessary();
+		}
+		Iterator(const Iterator& other)
+		    : m_it(other.m_it), m_end(other.m_end), m_obj(other.m_obj)
+		{
+			create_and_forward_if_necessary();
+		}
+		Iterator(Iterator&& other)
+		    : m_it(other.m_it), m_end(other.m_end), m_obj(other.m_obj)
+		{
+			create_and_forward_if_necessary();
+		}
+		~Iterator()
+		{
+			SG_UNREF(m_obj);
+		}
+
+		Iterator& operator=(const Iterator&) = delete;
+		Iterator& operator++()
+		{
+			SG_UNREF(m_obj);
+			do
+			{
+				++m_it;
+
+				if (m_it == m_end)
+					return *this;
+
+				m_obj = create_obj();
+			} while (!m_obj);
+
+			return *this;
+		}
+		CSGObject* operator*()
+		{
+			return m_obj;
+		}
+		bool operator!=(const Iterator& other) const
+		{
+			return this->m_it != other.m_it;
+		}
+
+	private:
+		CSGObject* create_obj()
+		{
+			auto obj = create((*m_it).c_str(), sg_primitive_type<T>());
+			SG_REF(obj);
+			return obj;
+		}
+
+		void create_and_forward_if_necessary()
+		{
+			if (m_it == m_end)
+				return;
+
+			m_obj = create_obj();
+			while (!m_obj && m_it != m_end)
+			{
+				this->operator++();
+			}
+		}
+
+		std::set<std::string>::iterator m_it;
+		std::set<std::string>::iterator m_end;
+		CSGObject* m_obj;
+	};
+
+	Iterator begin() const
+	{
+		return Iterator(m_class_names.begin(), m_class_names.end());
+	}
+	Iterator end() const
+	{
+		return Iterator(m_class_names.end(), m_class_names.end());
+	}
+
+	std::set<std::string> m_class_names;
+};
+
+TYPED_TEST(SGObjectAll, sg_object_iterator)
+{
+	for (auto obj : sg_object_iterator<TypeParam>())
+	{
+		SCOPED_TRACE(obj->get_name());
+		ASSERT_NE(obj, nullptr);
+		ASSERT_EQ(obj->ref_count(), 1);
+	}
+}
+
+std::set<std::string> sg_object_all_ignores = {"Set", "ParseBuffer",
+                                               "TreeMachine"};
+
+TYPED_TEST(SGObjectAll, clone_basic)
+{
+	for (auto obj : sg_object_iterator<TypeParam>(sg_object_all_ignores))
+	{
+		CSGObject* clone = nullptr;
+		try
+		{
+			clone = obj->clone();
+		}
+		catch (...)
+		{
+		}
+
+		ASSERT_NE(clone, nullptr);
+		EXPECT_EQ(clone->ref_count(), 1);
+		EXPECT_EQ(std::string(clone->get_name()), std::string(obj->get_name()));
+
+		SG_UNREF(clone);
+	}
+}
+
+TYPED_TEST(SGObjectAll, clone_equals)
```

Comment:
This replaces the old clone equals tests. Can obviously also be done for the serialization tests.